File: internal/parser/renderer.go
package parser

import (
	"fmt"
)

func Renderer(token node) string {
	if text, ok := token.(text); ok {
		return text.Value
	}

	element, ok := token.(element)
	if !ok {
		panic("[ERROR] Reached unreachable branch in renderer")
	}

	attributesList := ""
	if element.Properties != nil {
		for key, val := range element.Properties {
			attributesList += fmt.Sprintf(` %s="%s"`, key, val)
		}
	}

	contents := ""
	childrenLen := len(element.Children)

	if childrenLen == 0 {
		return fmt.Sprintf("<%s%s />", element.TagName, attributesList)
	}

	for _, item := range element.Children {
		contents += Renderer(item)
	}
	return fmt.Sprintf("<%s%s>%s</%s>", element.TagName, attributesList, contents, element.TagName)
}

File: crates/znak/src/parser/renderer.rs
use crate::parser::types::Node;

pub(crate) fn renderer(token: Node) -> String {
    match token {
        Node::Text(text) => text,
        Node::Element {
            tag_name,
            children,
            properties,
        } => {
            let mut attributes_list = String::new();
            for (key, val) in properties {
                attributes_list += format!(" {key}=\"{val}\"").as_str();
            }

            let mut contents = String::new();
            let children_len = children.len();

            if children_len == 0 {
                return format!("<{}{} />", tag_name, attributes_list);
            }

            for item in children {
                contents += renderer(item).as_str();
            }

            format!(
                "<{}{}>{}</{}>",
                tag_name, attributes_list, contents, tag_name
            )
        }
    }
}
